src/sentry/data_export/tasks.py

```diff
@@ -26,152 +37,198 @@ def assemble_download(
-    is_completed = False
-    while not is_completed:
-        offset = batch_size * iteration
-        next_offset = batch_size * (iteration + 1)
-        is_exceeding_limit = export_limit and export_limit < next_offset
-        raw_data_unicode = processor.data_fn(offset=offset, limit=batch_size)["data"]
-        # TODO(python3): Remove next line once the 'csv' module has been updated to Python 3
-        # See associated comment in './utils.py'
-        raw_data = convert_to_utf8(raw_data_unicode)
-        raw_data = processor.handle_fields(raw_data)
-        if is_exceeding_limit:
-            # Since the next offset will pass the export_limit, just write the remainder
-            writer.writerows(raw_data[: export_limit % batch_size])
-        else:
-            writer.writerows(raw_data)
-            iteration += 1
-        # If there are no returned results, or we've passed the export_limit, stop iterating
-        is_completed = len(raw_data) == 0 or is_exceeding_limit
+        bytes_offset += blob.size
+
+        # there is a maximum file size allowed, so we need to make sure we don't exceed it
+        if bytes_written + bytes_offset >= MAX_FILE_SIZE:
+            transaction.set_rollback(True)
+            return 0
+
 
+@instrumented_task(name="sentry.data_export.tasks.merge_blobs", queue="data_export")
+def merge_export_blobs(data_export_id, **kwargs):
+    try:
+        data_export = ExportedData.objects.get(id=data_export_id)
+    except ExportedData.DoesNotExist as error:
+        capture_exception(error)
+        return
 
-def create_writer(file, fields):
-    writer = csv.DictWriter(file, fields, extrasaction="ignore")
-    writer.writeheader()
-    return writer
+    # adapted from `putfile` in  `src/sentry/models/file.py`
+    try:
+        with transaction.atomic():
+            file = File.objects.create(
+                name=data_export.file_name, type="export.csv", headers={"Content-Type": "text/csv"},
+            )
+            size = 0
+            file_checksum = sha1(b"")
+
+            for export_blob in ExportedDataBlob.objects.filter(data_export=data_export).order_by(
+                "offset"
+            ):
+                blob = export_blob.blob
+                FileBlobIndex.objects.create(file=file, blob=blob, offset=size)
+                size += blob.size
+                blob_checksum = sha1(b"")
+
+                for chunk in blob.getfile().chunks():
+                    blob_checksum.update(chunk)
+                    file_checksum.update(chunk)
+
+                if blob.checksum != blob_checksum.hexdigest():
+                    raise AssembleChecksumMismatch("Checksum mismatch")
+
+            file.size = size
+            file.checksum = file_checksum.hexdigest()
+            data_export.finalize_upload(file=file)
+
+            logger.info("dataexport.end", extra={"data_export_id": data_export_id})
+            metrics.incr("dataexport.end", sample_rate=1.0)
+    except BaseException as error:
```

Comment:
Should you let errors related to restarts bubble up to celery so it can retry this task?

Reply:
Are retries handled automatically by celery? I'm currently looking at how that works because in my local environment, I haven't seen any errored tasks restart.

Reply:
You might need to set `max_retries` on the task, not sure what the defaults are.

Reply:
Looking into `max_retries` and `@retry` right now.